pilot/pkg/serviceregistry/kube/controller/ambientindex.go

```diff
@@ -387,45 +385,24 @@ func (c *Controller) setupIndex() *AmbientIndexImpl {
 		waypoints:                   map[model.WaypointScope]*workloadapi.GatewayAddress{},
 		serviceByAddr:               map[networkAddress]*model.ServiceInfo{},
 		serviceByNamespacedHostname: map[string]*model.ServiceInfo{},
+		servicesMap:                 map[types.NamespacedName]*apiv1alpha3.ServiceEntry{},
 	}
 
-	podHandler := cache.ResourceEventHandlerFuncs{
-		AddFunc: func(obj any) {
-			idx.mu.Lock()
-			defer idx.mu.Unlock()
-			updates := idx.handlePod(nil, obj, false, c)
-			if len(updates) > 0 {
-				c.opts.XDSUpdater.ConfigUpdate(&model.PushRequest{
-					ConfigsUpdated: updates,
-					Reason:         model.NewReasonStats(model.AmbientUpdate),
-				})
-			}
-		},
-		UpdateFunc: func(oldObj, newObj any) {
-			idx.mu.Lock()
-			defer idx.mu.Unlock()
-			updates := idx.handlePod(oldObj, newObj, false, c)
-			if len(updates) > 0 {
-				c.opts.XDSUpdater.ConfigUpdate(&model.PushRequest{
-					ConfigsUpdated: updates,
-					Reason:         model.NewReasonStats(model.AmbientUpdate),
-				})
-			}
-		},
-		DeleteFunc: func(obj any) {
-			idx.mu.Lock()
-			defer idx.mu.Unlock()
-			updates := idx.handlePod(nil, obj, true, c)
-			if len(updates) > 0 {
-				c.opts.XDSUpdater.ConfigUpdate(&model.PushRequest{
-					ConfigsUpdated: updates,
-					Reason:         model.NewReasonStats(model.AmbientUpdate),
-				})
-			}
-		},
+	podHandler := func(old, pod *v1.Pod, ev model.Event) error {
+		log.Infof("PodHandler pod %s/%s, event %v", pod.Namespace, pod.Name, ev)
+		idx.mu.Lock()
+		defer idx.mu.Unlock()
+		updates := idx.handlePod(nil, pod, ev, c)
+		if len(updates) > 0 {
+			c.opts.XDSUpdater.ConfigUpdate(&model.PushRequest{
+				ConfigsUpdated: updates,
+				Reason:         model.NewReasonStats(model.AmbientUpdate),
+			})
+		}
+		return nil
 	}
 
-	c.podsClient.AddEventHandler(podHandler)
+	registerHandlers[*v1.Pod](c, c.podsClient, "Pods", podHandler, nil)
```

Comment:
I think we are going to get duplicate metrics (from `incrementEvent`) since we have 2 pod handlers

Reply:
ACK